src/entities/swapRefactorDraft.ts

```diff
@@ -0,0 +1,376 @@
+import {
+    abs,
+    BALANCER_QUERIES,
+    DEFAULT_USERDATA,
+    DEFAULT_FUND_MANAGMENT,
+    ZERO_ADDRESS,
+    NATIVE_ADDRESS,
+    MathSol,
+} from '../utils';
+import {
+    Address,
+    createPublicClient,
+    encodeFunctionData,
+    getContract,
+    http,
+} from 'viem';
+import { balancerQueriesAbi } from '../abi';
+import { PriceImpactAmount } from './priceImpactAmount';
+import cloneDeep from 'lodash.clonedeep';
+import { Token } from './token';
+import { MinimalToken } from '..';
+
+/**
+API Returns list of GqlSorPath:
+
+type GqlSorPath {
+   pools: [String]! --- note can this be address?
+   tokens: [Token]!
+   outputAmountRaw: String!
+   inputAmountRaw: String!
+}
+
+type Token {
+    address: String!
+    decimals: Int!
+}
+*/
+
+type TokenApi = Omit<MinimalToken, 'index'>;
+
+type Path = {
+    pools: Address[];
+    tokens: TokenApi[];
+    outputAmountRaw: bigint;
+    inputAmountRaw: bigint;
+};
+
+class PathWithAmount {
+    public readonly pools: Address[];
+    public readonly tokens: TokenApi[];
+    public readonly outputAmount: TokenAmount;
+    public readonly inputAmount: TokenAmount;
+
+    public constructor(
+        chainId: number,
+        tokens: TokenApi[],
+        pools: Address[],
+        inputAmountRaw: bigint,
+        outputAmountRaw: bigint,
+    ) {
+        if (pools.length === 0 || tokens.length < 2) {
+            throw new Error(
+                'Invalid path: must contain at least 1 pool and 2 tokens.',
+            );
+        }
+        if (tokens.length !== pools.length + 1) {
+            throw new Error(
+                'Invalid path: tokens length must equal pools length + 1',
+            );
+        }
+
+        const tokenIn = new Token(
+            chainId,
+            tokens[0].address,
+            tokens[0].decimals,
+        );
+        const tokenOut = new Token(
+            chainId,
+            tokens[tokens.length - 1].address,
+            tokens[tokens.length - 1].decimals,
+        );
+        this.pools = pools;
+        this.tokens = tokens;
+        this.inputAmount = TokenAmount.fromRawAmount(tokenIn, inputAmountRaw);
+        this.outputAmount = TokenAmount.fromRawAmount(
+            tokenOut,
+            outputAmountRaw,
+        );
+    }
+}
+
+// A Swap can be a single or multiple paths
+export class Swap {
+    public constructor({
+        chainId,
+        paths,
+        swapKind,
+    }: { chainId: number; paths: Path[]; swapKind: SwapKind }) {
+        if (paths.length === 0)
+            throw new Error('Invalid swap: must contain at least 1 path.');
+
+        this.paths = paths.map(
+            (p) =>
+                new PathWithAmount(
+                    chainId,
+                    p.tokens,
+                    p.pools,
+                    p.inputAmountRaw,
+                    p.outputAmountRaw,
+                ),
+        );
+
+        // paths with immutable pool balances
+        this.pathsImmutable = cloneDeep(this.paths);
```

Comment:
This was needed in the context of the previous PathWithAmount that had mutating balances. Since this is no longer the case, you should be able to remove it.